From ddb021dce4f218ad269281c930122ad4dd2ce780 Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Sat, 19 Sep 2015 08:09:42 +0200
Subject: [PATCH 22/35] Enable lockdown integrity by default

It can be disabled by passing lockdown_disable on the kernel
command-line, only if Secure Boot is not enabled.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>

[ Corentin: rebase from XS-9 kernel (v6.6.98 based) to UEK8 kernel (v6.12 based):
- 260017f31a8c `("lsm: use default hook return value in call_int_hook()") simplified the `call_int_hook`
macro, so align our usage in `security_lock_kernel_down()`]
---
 arch/x86/kernel/setup.c       | 18 ++++++++++++++++++
 arch/x86/xen/efi.c            |  1 +
 include/linux/lsm_hook_defs.h |  1 +
 include/linux/security.h      |  5 +++++
 security/lockdown/lockdown.c  |  1 +
 security/security.c           |  6 ++++++
 6 files changed, 32 insertions(+)

diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index cb405c7edbf7..388be8258eab 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -714,6 +714,17 @@ static void __init x86_report_nx(void)
 	}
 }
 
+static bool lockdown_disabled;
+
+static int __init lockdown_disable_param(char *str)
+{
+	lockdown_disabled = true;
+
+	return 0;
+}
+
+early_param("lockdown_disable", lockdown_disable_param);
+
 /*
  * Determine if we were loaded by an EFI loader.  If so, then we have also been
  * passed the efi memmap, systab, etc., so we should use these data structures
@@ -906,6 +917,13 @@ void __init setup_arch(char **cmdline_p)
 		efi_init();
 
 	efi_set_secure_boot(boot_params.secure_boot);
+	if (boot_params.secure_boot == efi_secureboot_mode_enabled)
+		security_lock_kernel_down("EFI Secure Boot mode",
+					  LOCKDOWN_INTEGRITY_MAX);
+	else if (!lockdown_disabled)
+		security_lock_kernel_down("vendor policy",
+					  LOCKDOWN_INTEGRITY_MAX);
+
 	reserve_ibft_region();
 	x86_init.resources.dmi_setup();
 
diff --git a/arch/x86/xen/efi.c b/arch/x86/xen/efi.c
index 7250d0e0e1a9..eacaa2f27e34 100644
--- a/arch/x86/xen/efi.c
+++ b/arch/x86/xen/efi.c
@@ -7,6 +7,7 @@
 #include <linux/efi.h>
 #include <linux/init.h>
 #include <linux/string.h>
+#include <linux/security.h>
 
 #include <xen/xen.h>
 #include <xen/xen-ops.h>
diff --git a/include/linux/lsm_hook_defs.h b/include/linux/lsm_hook_defs.h
index 9eca013aa5e1..23041e1201d4 100644
--- a/include/linux/lsm_hook_defs.h
+++ b/include/linux/lsm_hook_defs.h
@@ -438,6 +438,7 @@ LSM_HOOK(int, 0, bpf_token_capable, const struct bpf_token *token, int cap)
 #endif /* CONFIG_BPF_SYSCALL */
 
 LSM_HOOK(int, 0, locked_down, enum lockdown_reason what)
+LSM_HOOK(int, 0, lock_kernel_down, const char *where, enum lockdown_reason level)
 
 #ifdef CONFIG_PERF_EVENTS
 LSM_HOOK(int, 0, perf_event_open, struct perf_event_attr *attr, int type)
diff --git a/include/linux/security.h b/include/linux/security.h
index 2ec8f3014757..94e63443b798 100644
--- a/include/linux/security.h
+++ b/include/linux/security.h
@@ -529,6 +529,7 @@ void security_bdev_free(struct block_device *bdev);
 int security_bdev_setintegrity(struct block_device *bdev,
 			       enum lsm_integrity_type type, const void *value,
 			       size_t size);
+int security_lock_kernel_down(const char *where, enum lockdown_reason level);
 #else /* CONFIG_SECURITY */
 
 static inline int call_blocking_lsm_notifier(enum lsm_event event, void *data)
@@ -1527,6 +1528,10 @@ static inline int security_bdev_setintegrity(struct block_device *bdev,
 	return 0;
 }
 
+static inline int security_lock_kernel_down(const char *where, enum lockdown_reason level)
+{
+	return 0;
+}
 #endif	/* CONFIG_SECURITY */
 
 #if defined(CONFIG_SECURITY) && defined(CONFIG_WATCH_QUEUE)
diff --git a/security/lockdown/lockdown.c b/security/lockdown/lockdown.c
index 8347359ea2f1..7557405619eb 100644
--- a/security/lockdown/lockdown.c
+++ b/security/lockdown/lockdown.c
@@ -75,6 +75,7 @@ static int lockdown_is_locked_down(enum lockdown_reason what)
 
 static struct security_hook_list lockdown_hooks[] __ro_after_init = {
 	LSM_HOOK_INIT(locked_down, lockdown_is_locked_down),
+	LSM_HOOK_INIT(lock_kernel_down, lock_kernel_down),
 };
 
 static const struct lsm_id lockdown_lsmid = {
diff --git a/security/security.c b/security/security.c
index c5981e558bc2..76e01d81ee29 100644
--- a/security/security.c
+++ b/security/security.c
@@ -5769,6 +5769,12 @@ int security_locked_down(enum lockdown_reason what)
 }
 EXPORT_SYMBOL(security_locked_down);
 
+int security_lock_kernel_down(const char *where, enum lockdown_reason level)
+{
+	return call_int_hook(lock_kernel_down, where, level);
+}
+EXPORT_SYMBOL(security_lock_kernel_down);
+
 /**
  * security_bdev_alloc() - Allocate a block device LSM blob
  * @bdev: block device
-- 
2.47.3

